source/extensions/upstreams/http/tcp/upstream_request.h

```diff
@@ -26,6 +26,10 @@ class TcpConnPool : public Router::GenericConnPool, public Envoy::Tcp::Connectio
               Upstream::ResourcePriority priority, Upstream::LoadBalancerContext* ctx) {
     conn_pool_data_ = thread_local_cluster.tcpConnPool(priority, ctx);
   }
+  ~TcpConnPool() override {
+    cancelAnyPendingStream();
+    ENVOY_BUG(upstream_handle_ == nullptr, "upstream_handle not null");
```

Comment:
isn't a no-op to call the ENVOY_BUG after the cancel? I think it needs to be called before to be useful
/wait

Reply:
🤦

Reply:
argh also just noticed cancelAnyPendingStream is virtual.
Can you move the code to a non-virtual cancelAnyPendingStreamImpl and call that here, to avoid any issues with virtual code from destructor?
Hopefully the last pass!